Warn the user if they're specifying protocols that won't be used

amcasey opened the pull request:
# Warn the user if they're specifying protocols that won't be used

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

## Description

If TLS is disabled and HTTP/1.x is available, HTTP/2 and HTTP/3 will be disabled.  This is predictable for HTTP/3, which requires TLS, but may surprise users of HTTP/2 which can be used without TLS if HTTP/1.x is not available.
    
Fixes #43837

amcasey committed:
Warn the user if they're specifying protocols that won't be used

If TLS is disabled and HTTP/1.x is available, HTTP/2 and HTTP/3 will be disabled.  This is predictable for HTTP/3, which requires TLS, but may surprise users of HTTP/2 which can be used without TLS if HTTP/1.x is not available.

Fixes #43837

amcasey committed:
Preserve ProtocolsSetExplicitly when cloning ListenOptions

amcasey committed:
Include endpoint in log message to reduce duplication

amcasey committed:
Add tests

amcasey commented:
Notes for reviewers

1. I did my best to follow local coding conventions, but please point out any I missed.
2. Please point out anywhere I'm using C# in an old-fashioned way.
3. This was mostly an exercise as the OP has already closed their issue.  Feel free to reject (after suggesting corrections).

JamesNK commented on src/Servers/Kestrel/Core/src/CoreStrings.resx:
```diff
@@ -719,4 +719,10 @@ For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?l
+    <value>Disabling HTTP/2 for {endpoint} since TLS is disabled and HTTP/1.1 is available.</value>
+  </data>
+  <data name="Http3DisabledWithHttp1AndNoTls" xml:space="preserve">
+    <value>Disabling HTTP/3 for {endpoint} since TLS is disabled and HTTP/1.1 is available.</value>
```
```suggestion
    <value>HTTP/3 is not enabled for {endpoint}. HTTP/3 requires TLS. Connections to this endpoint will use HTTP/1.1.</value>
```

JamesNK commented on src/Servers/Kestrel/Core/src/CoreStrings.resx:
```diff
@@ -719,4 +719,10 @@ For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?l
-</root>
\ No newline at end of file
+  <data name="Http2DisabledWithHttp1AndNoTls" xml:space="preserve">
+    <value>Disabling HTTP/2 for {endpoint} since TLS is disabled and HTTP/1.1 is available.</value>
```
```suggestion
    <value>HTTP/2 is not enabled for {endpoint}. The endpoint is configured to use HTTP/1.1 and HTTP/2, but TLS is not enabled. HTTP/2 requires TLS application protocol negotiation. Connections to this endpoint will use HTTP/1.1.</value>
```

JamesNK reviewed (commented):
I think the logs need some explanation added to help the dev

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs:
```diff
@@ -166,6 +166,18 @@ async Task OnBind(ListenOptions options, CancellationToken onBindCancellationTok
+                        {
+                            if (hasHttp2)
+                            {
+                                Trace.LogWarning(CoreStrings.Http2DisabledWithHttp1AndNoTls, options.EndPoint);
```
Usually, logging is done using defined logs. See KestrelTrace. I'm not sure why logging in this file is using LogWarning/LogInfo/etc directly.

amcasey replied:
I'm not sure what "defined logs" are - maybe they're more structured or have corresponding ETW events?

amcasey replied:
Do we localize those messages?  If so, how?

JamesNK replied:
See KestrelTrace - https://github.com/dotnet/aspnetcore/blob/9114f148b0ec923ebad046207598fcfa092f9d4e/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.General.cs

Most logging hangs off it.

We put log messages in resource files, but as far as I know, we don't localize.

amcasey replied:
To clarify, I understand that you're suggesting I use `KestrelTrace` and `LoggerMessageAttribute`, but I'm wondering what all that implies and whether "defined logs" is a more widely used term I should be aware of.

BrennanConroy replied:
"defined logs" refers to the `LoggerMessage.Define` way of logging
https://learn.microsoft.com/dotnet/core/extensions/high-performance-logging#define-a-logger-message

davidfowl replied:
We don't use "trace." anywhere, we use Microsoft.Extensions.Logging. @JamesNK is pointing you at the existing logs so you can add more to that.

Tratcher replied:
Oh, Trace here is KestrelTrace...

amcasey replied:
Added the messages to `KestrelTrace.GeneralLog`.

amcasey commented:
Force pushed a trivial rebase on main.  In particular, I have yet to adopt the defined logging approach.

halter73 commented on src/Servers/Kestrel/Core/src/LocalhostListenOptions.cs:
```diff
@@ -68,14 +68,20 @@ internal ListenOptions Clone(IPAddress address)
+        if (ProtocolsSetExplicitly)
+        {
+            options.Protocols = Protocols;
+        }
```
Would any of the existing tests fail if `Protocols` or `ProtocolsSetExplicitly` wasn't properly set here?

I think we should add a test for this `Clone` method similar to the one we have for `SniOptionsSelector.CloneSslOptions`. It uses reflection to put all the `SslOptions` property names in a list and makes sure we've checked all the properites in the lst when we're done testing the equality of all the properties after cloning.

https://github.com/dotnet/aspnetcore/blob/600eb9aa53c052ec7327e2399744215dbe493a89/src/Servers/Kestrel/Core/test/SniOptionsSelectorTests.cs#L901-L905

As a nit, I think it also might be simpler to just set `_protocols` and everything else directly rather than add conditional logic.

```csharp
        var options = new ListenOptions(new IPEndPoint(address, IPEndPoint!.Port))
        {
            KestrelServerOptions = KestrelServerOptions,
            // Set _protocols instead of Protocols to avoid side-effects
            _protocols = protocols,
            ProtocolsSetExplicitly = ProtocolsSetExplicitly,
// ...
```

amcasey replied:
Re test: Yep, I should add one.
Re nit: I didn't love the idea of making `_protocols` internal, but I'll follow the local convention.

amcasey committed:
Test LocalhostListenOptions.Clone

amcasey committed:
Move LocalhostListenOptions into base type

...and make the fields it accesses private.  Eliminate a now-unnecessary conditional.

amcasey committed:
Use defined logging

amcasey commented:
I believe this is ready to go, but I'd appreciate opinions on whether we actually want to merge this - I mostly did it as an exercise because the original bug was already withdrawn by the author.

amcasey commented on src/Servers/Kestrel/Core/src/CoreStrings.resx:
```diff
@@ -719,4 +719,4 @@ For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?l
   </data>
-</root>
\ No newline at end of file
+</root>
```
If I make another push, I'll clean up this newline - otherwise, it doesn't seem worth re-running CI.

amcasey commented on src/Servers/Kestrel/Core/test/KestrelServerTests.cs:
```diff
@@ -145,6 +145,64 @@ public void StartWarnsWhenIgnoringIServerAddressesFeature(string ignoredAddress)
+        using (var server = CreateServer(kestrelOptions, testLogger))
+        {
+            StartDummyApplication(server);
+            return (testLogger.Messages.Where(log => log.EventId == 64), testLogger.Messages.Where(log => log.EventId == 65));
```
I don't love these magic numbers (though they arguably can't change without breaking back-compat).  Would it be worthwhile to define constants?

Tratcher replied:
This is tolerable since it's test code and the test will fail immediately if it ever changes.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.General.cs:
```diff
@@ -82,6 +93,12 @@ private static partial class GeneralLog
         public static partial void InvalidResponseHeaderRemoved(ILogger logger);
 
-        // Highest shared ID is 63. New consecutive IDs start at 64
+        [LoggerMessage(64, LogLevel.Warning, "HTTP/2 is not enabled for {Endpoint}. The endpoint is configured to use HTTP/1.1 and HTTP/2, but TLS is not enabled. HTTP/2 requires TLS application protocol negotiation. Connections to this endpoint will use HTTP/1.1.", EventName = "Http2DisabledWithHttp1AndNoTls")]
+        public static partial void Http2DisabledWithHttp1AndNoTls(ILogger logger, EndPoint endPoint);
+
+        [LoggerMessage(65, LogLevel.Information, "HTTP/3 is not enabled for {Endpoint}. HTTP/3 requires TLS. Connections to this endpoint will use HTTP/1.1.", EventName = "Http3DisabledWithHttp1AndNoTls")]
+        public static partial void Http3DisabledWithHttp1AndNoTls(ILogger logger, EndPoint endPoint);
```
Why are the log levels different? The conditions are the same. Use Warning for both?

amcasey replied:
I went with Info for http/3 because it _never_ works without TLS, whereas I understand http/2 can work without TLS with some sort of pre-arranged setup (and I was tacitly assuming that I was the only one who didn't know that).

Tratcher replied:
That seems backwards. With that logic I'd expect this to be an Error 😁. Let's stick with Warning for both.

amcasey replied:
Will do, but note that there _is_ an error just below my change (when 1.1 isn't available as a fallback).

Tratcher reviewed (approved):
> I believe this is ready to go, but I'd appreciate opinions on whether we actually want to merge this - I mostly did it as an exercise because the original bug was already withdrawn by the author.

Let's merge it, it still has value.

amcasey committed:
Address PR feedback

Drop new linebreak.
Use the same trace level for both messages.

adityamandaleeka commented:
@mgravell just mentioned that this is hit often, especially in gRPC scenarios, so let's mention it in the blog :)
